### PR TITLE
Rework Hotkey Mechanism

### PR DESCRIPTION
**Pull request by DarkLordZach**
Was playing and wanted to add more keyboard shortcuts.

Added Shortcuts (Default Keys in Parentheses):
- Install CIA
- Exit (Alt-F4)
- Pause
- Stop
- Configure (,)
- Toggle Speed Limit (M)
- Toggle Emulation (Space)

Let me know if you want me to add more.

-------

Also takeover of #2773.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3460)

### CODE REVIEW

**Comment by LogPod** [comment]
Can you map them or are they hard coded?
and i would like a toggle audio stretching hotkey if that's fine with you.

**Comment by DarkLordZach** [comment]
Okay, and they are add just like citra's other hotkeys, so you can alter them using the config file and maybe gui? There is a place in settings that lists them but I haven't been able to find out how to change in the gui.

**Comment by DarkLordZach** [comment]
@LogPod added audio stretching with no default in commit cfec212c4f48753764b408482b83a3bdf3889eaa

**Comment by adityaruplaha** [comment]
This should close #3417

**Comment by adityaruplaha** [comment]
LGTM

**Comment by adityaruplaha** [comment]
Imo, "," is not a good key to map Configure to, please consider Tab. But this is trivial, g only my personal opinion.

**Comment by wwylele** [comment]
I agree that some of this hard-coded binding are bad because users may want to bind them to game input (for example I use M as Select and Space as Start).

**Comment by Hexagon12** [comment]
The hotkeys should be configurable with a controller too.

Also, you might want to look through the comments on this PR: #3340, since the whole hotkey configuration system needs to be reworked (best in a seperate tab).

**Comment by BreadFish64** [comment]
I would look at #2773 and take some inspiration to make hotkeys configurable.

File: src/citra_qt/main.cpp [hunk 1]
```diff
@@ -277,7 +277,12 @@ void GMainWindow::InitializeRecentFileMenuActions() {
 
 void GMainWindow::InitializeHotkeys() {
     RegisterHotkey("Main Window", "Load File", QKeySequence::Open);
+    RegisterHotkey("Main Window", "Install CIA");
+    RegisterHotkey("Main Window", "Exit", QKeySequence(Qt::ALT + Qt::Key_F4));
```
Comment (B3n30) [inline src/citra_qt/main.cpp]: Should be QKeySequence::Quit for other os's

File: src/citra_qt/main.cpp [hunk 2]
```diff
@@ -287,12 +292,25 @@ void GMainWindow::InitializeHotkeys() {
                    Qt::ApplicationShortcut);
     RegisterHotkey("Main Window", "Decrease Speed Limit", QKeySequence("-"),
                    Qt::ApplicationShortcut);
+    RegisterHotkey("Main Window", "Toggle Speed Limit", QKeySequence("M"), Qt::ApplicationShortcut);
+    RegisterHotkey("Main Window", "Toggle Emulation", QKeySequence(" "), Qt::ApplicationShortcut);
```
Comment (B3n30) [inline src/citra_qt/main.cpp]: I think if toggle is there Start Emulation/Pause Emulation isn't needed. Or if Start and Pause hotkey are available Toggle Emulation isn't needed

File: src/citra_qt/main.cpp [hunk 3]
```diff
@@ -287,12 +292,25 @@ void GMainWindow::InitializeHotkeys() {
                    Qt::ApplicationShortcut);
     RegisterHotkey("Main Window", "Decrease Speed Limit", QKeySequence("-"),
                    Qt::ApplicationShortcut);
+    RegisterHotkey("Main Window", "Toggle Speed Limit", QKeySequence("M"), Qt::ApplicationShortcut);
```
Comment (B3n30) [inline src/citra_qt/main.cpp]: "M" is default for "select" in citra. So it shouldn't be mapped to an hotkey

File: src/citra_qt/main.cpp (hunk 2 above)
Comment (B3n30) [inline src/citra_qt/main.cpp]: Space is default for "Start" in citra so it shouldn't be an hotkey

File: src/citra_qt/main.cpp [hunk 4]
```diff
@@ -277,7 +277,12 @@ void GMainWindow::InitializeRecentFileMenuActions() {
 
 void GMainWindow::InitializeHotkeys() {
     RegisterHotkey("Main Window", "Load File", QKeySequence::Open);
+    RegisterHotkey("Main Window", "Install CIA");
+    RegisterHotkey("Main Window", "Exit", QKeySequence(Qt::ALT + Qt::Key_F4));
     RegisterHotkey("Main Window", "Start Emulation");
+    RegisterHotkey("Main Window", "Pause Emulation");
+    RegisterHotkey("Main Window", "Stop Emulation");
+    RegisterHotkey("Main Window", "Configure Emulation", QKeySequence(","));
```
Comment (B3n30) [inline src/citra_qt/main.cpp]: I don't like "Configure Emulation" since it is the general "Configure"

File: src/citra_qt/main.cpp [hunk 5]
```diff
@@ -322,6 +340,20 @@ void GMainWindow::InitializeHotkeys() {
...
+        if (GMainWindow::ui.action_Start->isEnabled() && !GMainWindow::ui.action_Pause->isEnabled())
+            OnStartGame();
+        else if (!GMainWindow::ui.action_Start->isEnabled() &&
+                 GMainWindow::ui.action_Pause->isEnabled())
+            OnPauseGame();
+    });
+    connect(GetHotkey("Main Window", "Toggle Audio Stretching", this), &QShortcut::activated, this,
+            [&] {
+                Settings::values.enable_audio_stretching =
+                    !Settings::values.enable_audio_stretching;
```
Comment (B3n30) [inline src/citra_qt/main.cpp]: Did you check that this even works. Audio_Core only checks this value on init (Starting a game). So toggling that value doesn't change the behavior while a game is running

File: src/citra_qt/main.cpp [hunk 6]
```diff
@@ -897,7 +929,8 @@ void GMainWindow::OnMenuReportCompatibility() {
         QMessageBox::critical(
             this, tr("Missing Citra Account"),
             tr("In order to submit a game compatibility test case, you must link your Citra "
-               "account.<br><br/>To link your Citra account, go to Emulation &gt; Configuration &gt; "
+               "account.<br><br/>To link your Citra account, go to Emulation &gt; Configuration "
+               "&gt; "
```
Comment (B3n30) [inline src/citra_qt/main.cpp]: Why is this changed?

File: src/citra_qt/main.cpp [hunk 7]
```diff
@@ -1045,11 +1078,10 @@ void GMainWindow::UpdateStatusBar() {
 
     if (Settings::values.use_frame_limit) {
         emu_speed_label->setText(tr("Speed: %1% / %2%")
-                                    .arg(results.emulation_speed * 100.0, 0, 'f', 0)
-                                    .arg(Settings::values.frame_limit));
+                                     .arg(results.emulation_speed * 100.0, 0, 'f', 0)
```
Comment (B3n30) [inline src/citra_qt/main.cpp]: Why is this changed?

File: src/citra_qt/main.cpp [hunk 8]
```diff
@@ -1045,11 +1078,10 @@ void GMainWindow::UpdateStatusBar() {
...
+                                     .arg(Settings::values.frame_limit));
     } else {
-        emu_speed_label->setText(tr("Speed: %1%")
-                                    .arg(results.emulation_speed * 100.0, 0, 'f', 0));
+        emu_speed_label->setText(tr("Speed: %1%").arg(results.emulation_speed * 100.0, 0, 'f', 0));
```
Comment (B3n30) [inline src/citra_qt/main.cpp]: Why is this changed?

Comment (jroweboy) [reply src/citra_qt/main.cpp]: clang format changes are because we *still* haven't upgraded to 6 *cough cough* and this is the style for clang format 3

File: src/citra_qt/main.cpp (hunk 7 above)
Comment (DarkLordZach) [reply src/citra_qt/main.cpp]: Clang-format

File: src/citra_qt/main.cpp (hunk 6 above)
Comment (DarkLordZach) [reply src/citra_qt/main.cpp]: Clang-format

File: src/citra_qt/main.cpp (hunk 2 above)
Comment (DarkLordZach) [reply src/citra_qt/main.cpp]: I added this because personally I would like to be able to start/stop with one key. This is the reason I created this PR in the first place.

**Comment by DarkLordZach** [comment]
@BreadFish64 @B3n30 I am going to complete the review changes and add things from #2773. Consider me taking it over.

File: src/citra_qt/main.cpp (hunk 5 above)
Comment (DarkLordZach) [reply src/citra_qt/main.cpp]: @B3n30 would calling `Settings::Apply()` work. It would seem to. tbh did not test this.

Comment (DarkLordZach) [reply src/citra_qt/main.cpp]: Can confirm Settings::Apply() fixes this

**Comment by jroweboy** [comment]
My apologies! We just pushed a clang format update, and it looks like you will need to rebase this PR off the latest master and fix any conflicts. With the new clang format update, its really easy for people to fix any clang format issues, simply run the clang format build target. See the contributors guide for more information or ask me on IRC or discord. Thanks!

**Comment by jroweboy** [comment]
My apologies for finally getting around to reviewing the code after its been here for so long! Citra is kinda understaffed right now, especially considering most of the devs prefer to write code instead of review it :)

Anyway, I read through the code, and I guess I'm fine with it. I'm not happy with how hotkeys work in citra at the moment; they are "centralized" with register hotkey, but it only does half of what it needs to do (it doesn't allow remapping through the ui, binding to controllers, or binding a shortcut and a menu action to the same hotkey) Since this code is simply an extension of whats already there, all its doing is putting more work on the person who will rewrite it later, but I think this is a fair tradeoff.

I would have liked to see all the hotkeys and connections moved into the hotkey file, as main.cpp is already bloated enough as it is. I'd be okay if it doesn't happen, but i'd be happy if it does happen ;D

If someone feels like cleaning this up once more (ie fix merge conflicts, and maybe moving the signal/slots for hotkeys to some other location if its not too much of a pain) I'll recheck the changes and very likely approve it.

In the meantime, I'm just going to close it, as its clear that once PRs make it to the second page, they become out of sight and out of mind, and no one remembers to review them. Once again, I apologize for not getting around to this until just now.
